{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import division\n",
    "import glob, os, gc\n",
    "import os.path\n",
    "import csv\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import offsetbox\n",
    "from time import time\n",
    "from subprocess import (call, Popen, PIPE)\n",
    "from itertools import product\n",
    "from sklearn import (manifold, datasets, decomposition, ensemble, discriminant_analysis, random_projection)\n",
    "from sklearn.decomposition import (PCA, RandomizedPCA)\n",
    "from sklearn.datasets import fetch_mldata\n",
    "from sklearn.utils import shuffle\n",
    "from IPython.display import Image\n",
    "from PIL import Image\n",
    "from IPython.display import Image as IPImage\n",
    "import shutil\n",
    "import pybedtools\n",
    "import pysam\n",
    "import re\n",
    "import xml.etree.ElementTree as ET\n",
    "import time\n",
    "\n",
    "\n",
    "\n",
    "##Path to Data\n",
    "Path_SLAV = \"/raid/LOG-G/SEQ_ARCHIVE/Patrick_Reed/BSMN/SLAV_Data/\" \n",
    "Path_Common = \"/raid/LOG-G/SEQ_ARCHIVE/Patrick_Reed/BSMN/Common_Experiment/\"\n",
    "narrowpeak = \"-ready_peaks.narrowPeak\"\n",
    "peaks_merged = \"_peaksMerged.txt\"\n",
    "peaks_merged_bed = \"_peaksMerged.bed\"\n",
    "peaks_correct_bed = \"_peaksCorrect.bed\"\n",
    "peakregions_sml = \".peakregions_sml\"\n",
    "peakregions_lrg = \".peakregions_lrg\"\n",
    "peaks_correct_data = \"_peaksCorrect.data\"\n",
    "peaks_L1HS_bedgraph = \"_peaks_L1HS_mapped.bedgraph\"\n",
    "loci_sml = \".loci_sml\"\n",
    "loci_lrg = \".loci_lrg\"\n",
    "overlap = \"_overlap_\"\n",
    "overlap_sml = \"_overlap_sml_\"\n",
    "overlap_lrg = \"_overlap_lrg_\"\n",
    "bam = \"-ready.bam\"\n",
    "bed = \".bed\"\n",
    "## rmask Paths \n",
    "L1HS = \"/raid/LOG-G/SEQ_ARCHIVE/Patrick_Reed/BSMN/rmask_L1HS_Final.bed\"\n",
    "L1PA2 = \"/raid/LOG-G/SEQ_ARCHIVE/Patrick_Reed/BSMN/rmask_L1PA2.bed\"\n",
    "L1PA3 = \"/raid/LOG-G/SEQ_ARCHIVE/Patrick_Reed/BSMN/rmask_L1PA3.bed\"\n",
    "L1PA4 = \"/raid/LOG-G/SEQ_ARCHIVE/Patrick_Reed/BSMN/rmask_L1PA4.bed\"\n",
    "L1PA5 = \"/raid/LOG-G/SEQ_ARCHIVE/Patrick_Reed/BSMN/rmask_L1PA5.bed\"\n",
    "L1PA2345 = \"/raid/LOG-G/SEQ_ARCHIVE/Patrick_Reed/BSMN/rmask_L1PA2345_Final.bed\"\n",
    "L1_Other = \"/raid/LOG-G/SEQ_ARCHIVE/Patrick_Reed/BSMN/rmask_L1_Other_Final.bed\"\n",
    "\n",
    "##SLAV Data\n",
    "Bulk_1571_Cerebellum = \"1571_cereb_BT_40_L3\"\n",
    "Bulk_1571_Hippocampus = \"1571_hippo_BT_41_L3\"\n",
    "Bulk_1846_Cerebellum = \"1846_cereb_BT_13_L3\"\n",
    "Bulk_1846_Cortex = \"1846_cortex_BT_71_L3\"\n",
    "Bulk_1846_Hippocampus = \"1846_hippo_BT_19_L3\"\n",
    "Bulk_1846_Liver = \"1846_liver_BT_22_L3\"\n",
    "Bulk_5125_Cortex = \"5125_cortex_BT_122_L3\"\n",
    "Bulk_5125_Hippocampus = \"5125_hippo_BT_139_L3\"\n",
    "Bulk_5125_Liver = \"5125_liver_BT_164_L3\"\n",
    "SC_1571_Hippo = [\"1571_hippo_SC_43_L3\",\"1571_hippo_SC_45_L3\",\"1571_hippo_SC_46_L3\",\"1571_hippo_SC_47_L3\",\"1571_hippo_SC_48_L3\",\"1571_hippo_SC_50_L3\",\"1571_hippo_SC_51_L3\",\"1571_hippo_SC_52_L3\",\"1571_hippo_SC_53_L3\",\"1571_hippo_SC_55_L3\",\"1571_hippo_SC_56_L3\",\"1571_hippo_SC_57_L3\",\"1571_hippo_SC_58_L3\",\"1571_hippo_SC_59_L3\",\"1571_hippo_SC_61_L3\",\"1571_hippo_SC_62_L3\",\"1571_hippo_SC_63_L3\",\"1571_hippo_SC_64_L3\"]\n",
    "SC_1846_Cortex = [\"1846_cortex_SC_72_L3\",\"1846_cortex_SC_73_L3\",\"1846_cortex_SC_74_L3\",\"1846_cortex_SC_75_L3\",\"1846_cortex_SC_78_L3\",\"1846_cortex_SC_79_L3\",\"1846_cortex_SC_80_L3\",\"1846_cortex_SC_81_L3\",\"1846_cortex_SC_82_L3\",\"1846_cortex_SC_83_L3\",\"1846_cortex_SC_84_L3\",\"1846_cortex_SC_85_L3\",\"1846_cortex_SC_86_L3\"]\n",
    "SC_1846_Hippo = [\"1846_hippo_SC_100_L3\",\"1846_hippo_SC_101_L3\",\"1846_hippo_SC_102_L3\",\"1846_hippo_SC_103_L3\",\"1846_hippo_SC_104_L3\",\"1846_hippo_SC_105_L3\",\"1846_hippo_SC_106_L3\",\"1846_hippo_SC_107_L3\",\"1846_hippo_SC_108_L3\",\"1846_hippo_SC_109_L3\",\"1846_hippo_SC_110_L3\",\"1846_hippo_SC_111_L3\",\"1846_hippo_SC_112_L3\",\"1846_hippo_SC_113_L3\",\"1846_hippo_SC_88_L3\",\"1846_hippo_SC_89_L3\",\"1846_hippo_SC_90_L3\",\"1846_hippo_SC_91_L3\",\"1846_hippo_SC_92_L3\",\"1846_hippo_SC_93_L3\",\"1846_hippo_SC_94_L3\",\"1846_hippo_SC_95_L3\",\"1846_hippo_SC_99_L3\"]\n",
    "SC_5125_Cortex = [\"5125_cortex_SC_125_L3\",\"5125_cortex_SC_126_L3\",\"5125_cortex_SC_127_L3\",\"5125_cortex_SC_128_L3\",\"5125_cortex_SC_129_L3\",\"5125_cortex_SC_130_L3\",\"5125_cortex_SC_131_L3\",\"5125_cortex_SC_132_L3\",\"5125_cortex_SC_133_L3\",\"5125_cortex_SC_134_L3\",\"5125_cortex_SC_135_L3\",\"5125_cortex_SC_136_L3\",\"5125_cortex_SC_138_L3\"]\n",
    "SC_5125_Hippo = [\"5125_hippo_SC_140_L3\",\"5125_hippo_SC_141_L3\",\"5125_hippo_SC_142_L3\",\"5125_hippo_SC_143_L3\",\"5125_hippo_SC_144_L3\",\"5125_hippo_SC_145_L3\",\"5125_hippo_SC_147_L3\",\"5125_hippo_SC_149_L3\",\"5125_hippo_SC_150_L3\",\"5125_hippo_SC_151_L3\",\"5125_hippo_SC_152_L3\",\"5125_hippo_SC_153_L3\",\"5125_hippo_SC_154_L3\",\"5125_hippo_SC_155_L3\",\"5125_hippo_SC_156_L3\",\"5125_hippo_SC_157_L3\",\"5125_hippo_SC_158_L3\",\"5125_hippo_SC_159_L3\",\"5125_hippo_SC_160_L3\",\"5125_hippo_SC_161_L3\",\"5125_hippo_SC_162_L3\",\"5125_hippo_SC_163_L3\"]\n",
    "Data_Sets_SLAV = []\n",
    "Data_Sets_SLAV.append([SC_1571_Hippo,Bulk_1571_Hippocampus,Bulk_1571_Cerebellum])\n",
    "Data_Sets_SLAV.append([SC_1846_Cortex,Bulk_1846_Cortex,Bulk_1846_Liver])\n",
    "Data_Sets_SLAV.append([SC_1846_Hippo,Bulk_1846_Hippocampus,Bulk_1846_Liver])\n",
    "Data_Sets_SLAV.append([SC_5125_Cortex,Bulk_5125_Cortex,Bulk_5125_Liver])\n",
    "Data_Sets_SLAV.append([SC_5125_Hippo,Bulk_5125_Hippocampus,Bulk_5125_Liver])\n",
    "\n",
    "##Common Data\n",
    "Bulk_Brain_Common = \"04132016_mw_Bulk_cor\"\n",
    "Bulk_Fibro_Common = \"05252016_mw_Bulk_fib\"\n",
    "SC_MDA = [\"04132016_mw_1571_SC_B4_S48\",\"04132016_mw_1571_SC_D4_S49\",\"04132016_mw_L1B1_SC_A2_S43\",\"04132016_mw_L1B1_SC_C1_S45\",\"04132016_mw_L1B1_SC_C2_S46\",\"04132016_mw_L1B1_SC_D2_S50\",\"04132016_mw_L1B1_SC_E2_S51\",\"04132016_mw_L1B1_SC_E3_S52\",\"04132016_mw_L1B1_SC_E3_S52\",\"04132016_mw_L1B1_SC_F2_S53\",\"04132016_mw_L1B1_SC_G1_S54\",\"04132016_mw_L1B1_SC_H1_S55\",\"05252016_mw_L1B1_SC_B4_S47\"] \n",
    "SC_MALBAC = [\"2122_S1\",\"2178_S2\",\"2179_S3\",\"2180_S4\",\"2184_S5\",\"2186_S6\",\"2187_S7\",\"2188_S8\",\"2193_S9\",\"2196_S10\",\"2197_S11\",\"2198_S12\",\"2261_S13\",\"2263_S14\",\"2264_S15\",\"2265_S16\"] \n",
    "Data_Sets_Common = []\n",
    "Data_Sets_Common.append([SC_MALBAC,Bulk_Brain_Common,Bulk_Fibro_Common])\n",
    "Data_Sets_Common.append([SC_MDA,Bulk_Brain_Common,Bulk_Fibro_Common])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Percent L1HS Overlap: 60.9455958549\n",
      "Percent L1PA2 Overlap: 66.6870042709\n",
      "1571_hippo_SC_43_L3\n",
      "Percentage of Bulk L1HS detected in Cell: 94.3676939426\n",
      "Percentage of Bulk L1PA2 detected in Cell: 89.4784995425\n",
      "Total Aligned Reads: 8862220\n",
      "average aligned reads per L1HS: 1393.31531532\n",
      "average aligned reads per L1PA2: 871.66325835\n",
      "1571_hippo_SC_45_L3\n",
      "Percentage of Bulk L1HS detected in Cell: 92.6673751328\n",
      "Percentage of Bulk L1PA2 detected in Cell: 87.0692284233\n",
      "Total Aligned Reads: 12100577\n",
      "average aligned reads per L1HS: 1573.48394495\n",
      "average aligned reads per L1PA2: 1143.97723292\n",
      "1571_hippo_SC_46_L3\n",
      "Percentage of Bulk L1HS detected in Cell: 97.5557917109\n",
      "Percentage of Bulk L1PA2 detected in Cell: 80.6038426349\n",
      "Total Aligned Reads: 3382735\n",
      "average aligned reads per L1HS: 1933.49237473\n",
      "average aligned reads per L1PA2: 161.220204313\n",
      "1571_hippo_SC_47_L3\n",
      "Percentage of Bulk L1HS detected in Cell: 89.4792773645\n",
      "Percentage of Bulk L1PA2 detected in Cell: 83.043610857\n",
      "Total Aligned Reads: 9393969\n",
      "average aligned reads per L1HS: 1078.90617577\n",
      "average aligned reads per L1PA2: 940.533235402\n",
      "1571_hippo_SC_48_L3\n",
      "Percentage of Bulk L1HS detected in Cell: 95.7492029756\n",
      "Percentage of Bulk L1PA2 detected in Cell: 90.7593778591\n"
     ]
    }
   ],
   "source": [
    "for dset in Data_Sets_SLAV:  \n",
    "    Bulk_Brain_Peaks = os.path.join(Path_SLAV,  dset[1], \"peaks.bed\")\n",
    "    Bulk_NonBrain_Peaks = os.path.join(Path_SLAV,  dset[2], \"peaks.bed\")\n",
    "    a = pybedtools.BedTool(Bulk_Brain_Peaks)\n",
    "    b_L1HS = pybedtools.BedTool(L1HS)\n",
    "    L1HS_in_Bulk = b_L1HS.intersect(a)\n",
    "    output = \"Percent L1HS Overlap: \" + str((len(L1HS_in_Bulk) / len(b_L1HS))*100)\n",
    "    print output\n",
    "    b_L1PA2 = pybedtools.BedTool(L1PA2)\n",
    "    L1PA2_in_Bulk = b_L1PA2.intersect(a)\n",
    "    output = \"Percent L1PA2 Overlap: \" + str((len(L1PA2_in_Bulk) / len(b_L1PA2))*100)\n",
    "    print output\n",
    "    b_L1PA3 = pybedtools.BedTool(L1PA3)\n",
    "    L1PA3_in_Bulk = b_L1PA3.intersect(a)\n",
    "    b_L1PA4 = pybedtools.BedTool(L1PA4)\n",
    "    L1PA4_in_Bulk = b_L1PA4.intersect(a)\n",
    "    b_L1PA5 = pybedtools.BedTool(L1PA5)\n",
    "    L1PA5_in_Bulk = b_L1PA5.intersect(a)      \n",
    "    for cell in dset[0]:\n",
    "        print cell\n",
    "        os.chdir(os.path.join(Path_SLAV, cell))\n",
    "        name = os.path.join(Path_SLAV, cell, cell)\n",
    "        bamfile = os.path.join(Path_SLAV, cell, cell+bam)\n",
    "        Cell_Peaks = os.path.join(Path_SLAV, cell, \"peaks.bed\")\n",
    "        a = pybedtools.BedTool(Cell_Peaks)\n",
    "        L1HS_in_Cell = L1HS_in_Bulk.intersect(a)\n",
    "        output = \"Percentage of Bulk L1HS detected in Cell: \" + str((len(L1HS_in_Cell) / len(L1HS_in_Bulk))*100)\n",
    "        print output\n",
    "        L1PA2_in_Cell = L1PA2_in_Bulk.intersect(a)\n",
    "        output = \"Percentage of Bulk L1PA2 detected in Cell: \" + str((len(L1PA2_in_Cell) / len(L1PA2_in_Bulk))*100)\n",
    "        print output\n",
    "        L1HS_in_Cell_slop1K = L1HS_in_Cell.slop(b=1000, genome='hg19')\n",
    "        a_bam = pysam.AlignmentFile(os.path.join(Path_SLAV, cell, cell + bam), 'rb')\n",
    "        bam_reads=0\n",
    "        for read in a_bam.fetch(): bam_reads+=1\n",
    "        output = \"Total Aligned Reads: \" + str(bam_reads)  \n",
    "        print output\n",
    "        L1HS_reads=0\n",
    "        for interval in L1HS_in_Cell_slop1K:\n",
    "            for read in a_bam.fetch(str(interval[0]), int(interval[1]), int(interval[2])): L1HS_reads+=1 \n",
    "        output = \"average aligned reads per L1HS: \" + str((L1HS_reads / len(L1HS_in_Cell)))\n",
    "        print output\n",
    "        #output = \"Percent aligned reads that map to L1HS: \" + str((L1HS_reads / bam_reads)*100)\n",
    "        #print output\n",
    "        #output = \"average L1HS reads as percent of total: \" + str(((L1HS_reads / len(L1HS_in_Cell)) / bam_reads)*100)\n",
    "        #print output\n",
    "        L1PA2_in_Cell_slop1K = L1PA2_in_Cell.slop(b=1000, genome='hg19')\n",
    "        L1PA2_reads=0\n",
    "        for interval in L1PA2_in_Cell_slop1K:\n",
    "            for read in a_bam.fetch(str(interval[0]), int(interval[1]), int(interval[2])): L1PA2_reads+=1 \n",
    "        output = \"average aligned reads per L1PA2: \" + str((L1PA2_reads / len(L1PA2_in_Cell)))\n",
    "        print output\n",
    "        #output = \"Percent aligned reads that map to L1PA2: \" + str((L1PA2_reads / bam_reads)*100)\n",
    "        #print output\n",
    "        #output = \"average L1PA2 reads as percent of total: \" + str(((L1PA2_reads / len(L1PA2_in_Cell)) / bam_reads)*100)\n",
    "        #print output\n",
    "        \n",
    "        #percentage of reads on-target"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
